src/router.rs

```diff
@@ -12,49 +14,53 @@ struct MatcherKey(usize, Uuid);
 
 pub struct Router<'a> {
     schema: &'a Schema,
-    matchers: BTreeMap<MatcherKey, Expression>,
+    routes: BTreeMap<MatcherKey, Route>,
     pub fields: HashMap<String, usize>,
 }
 
 impl<'a> Router<'a> {
     pub fn new(schema: &'a Schema) -> Self {
         Self {
             schema,
-            matchers: BTreeMap::new(),
+            routes: BTreeMap::new(),
             fields: HashMap::new(),
         }
     }
 
+    pub fn convert_route(&mut self, ast: Expression) -> Route {
+        let mut route = Route::new();
+        ast.convert(&mut route);
+        route
+    }
+
     pub fn add_matcher(&mut self, priority: usize, uuid: Uuid, atc: &str) -> Result<(), String> {
         let key = MatcherKey(priority, uuid);
 
-        if self.matchers.contains_key(&key) {
+        if self.routes.contains_key(&key) {
             return Err("UUID already exists".to_string());
         }
 
-        let ast = parse(atc).map_err(|e| e.to_string())?;
-
-        ast.validate(self.schema)?;
-        ast.add_to_counter(&mut self.fields);
-
-        assert!(self.matchers.insert(key, ast).is_none());
+        let route = self.convert_route(parse(atc).map_err(|e| e.to_string())?);
+        route.validate(self.schema)?;
+        route.add_to_counter(&mut self.fields);
```

Comment:
Let's validate the AST first and then convert it to IR. IR strips out extra info which is necessary for the validator to produce user-friendly error messages.

Reply:
Fixed, validate the AST first. 
And here, at the moment, Route does not strip any info. Just in case, Expression will add more information in future.